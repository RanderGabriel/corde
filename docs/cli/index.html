<!doctype html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-alpha.64">
<title data-react-helmet="true">CLI | Corde</title><meta data-react-helmet="true" name="docsearch:version" content="2.0,latest"><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" property="og:title" content="CLI | Corde"><meta data-react-helmet="true" name="description" content="The corde command line runner has a number of useful options."><meta data-react-helmet="true" property="og:description" content="The corde command line runner has a number of useful options."><meta data-react-helmet="true" property="og:url" content="https://lucasgmagalhaes.github.io/corde/docs/cli"><link data-react-helmet="true" rel="shortcut icon" href="/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://lucasgmagalhaes.github.io/corde/docs/cli"><link rel="stylesheet" href="/styles.d4c24b89.css">
<link rel="preload" href="/styles.78702091.js" as="script">
<link rel="preload" href="/runtime~main.c80c9b84.js" as="script">
<link rel="preload" href="/main.a97df2ae.js" as="script">
<link rel="preload" href="/common.570c17a0.js" as="script">
<link rel="preload" href="/2.f7266800.js" as="script">
<link rel="preload" href="/29.0b1f2317.js" as="script">
<link rel="preload" href="/30.99c11695.js" as="script">
<link rel="preload" href="/78311276.2010c58c.js" as="script">
<link rel="preload" href="/17896441.364198ab.js" as="script">
<link rel="preload" href="/e6aa8ecf.28a4f4e9.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<nav class="navbar navbar--light navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><div aria-label="Navigation bar toggle" class="navbar__toggle" role="button" tabindex="0"><svg xmlns="http://www.w3.org/2000/svg" width="30" height="30" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></div><a class="navbar__brand" href="/"><img class="navbar__logo" src="/img/logo.png" alt="corde logo"><strong class="navbar__title">Corde</strong></a><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/docs/">Docs</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/lucasgmagalhaes/corde" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub</a><div class="react-toggle react-toggle--disabled displayOnlyInLargeViewport_2aTZ"><div class="react-toggle-track"><div class="react-toggle-track-check"><span class="toggle_BsTx">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_BsTx">ðŸŒž</span></div></div><div class="react-toggle-thumb"></div><input type="checkbox" disabled="" aria-label="Dark mode toggle" class="react-toggle-screenreader-only"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/"><img class="navbar__logo" src="/img/logo.png" alt="corde logo"><strong class="navbar__title">Corde</strong></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a aria-current="page" class="menu__link navbar__link--active" href="/docs/">Docs</a></li><li class="menu__list-item"><a href="https://github.com/lucasgmagalhaes/corde" target="_blank" rel="noopener noreferrer" class="menu__link">GitHub</a></li></ul></div></div></div></nav><div class="main-wrapper"><div class="docPage_2gpo"><div class="docSidebarContainer_3_JD" role="complementary"><div class="sidebar_2urC"><div class="menu menu--responsive menu_5FrY"><button aria-label="Open Menu" aria-haspopup="true" class="button button--secondary button--sm menu__button" type="button"><svg aria-label="Menu" class="sidebarMenuIcon_Dm3K" xmlns="http://www.w3.org/2000/svg" height="24" width="24" viewBox="0 0 32 32" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><ul class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!">Introduction</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/docs/">Getting Started</a></li><li class="menu__list-item"><a aria-current="page" class="menu__link menu__link--active active" tabindex="0" href="/docs/cli">CLI</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/docs/configurations">Configuring Corde</a></li></ul></li></ul></div></div></div><main class="docMainContainer_3EyW"><div class="container padding-vert--lg docItemWrapper_1EkI"><div class="row"><div class="col docItemCol_2ASc"><div class="docItemContainer_3QWW"><article><div><span class="badge badge--secondary">Version: 2.0</span></div><header><h1 class="docTitle_1Lrw">CLI</h1></header><div class="markdown"><p>The <code>corde</code> command line runner has a number of useful options.
Some of the options shown below can also be used together to run tests exactly the way you want.
Some one of Corde&#x27;s <a href="/docs/Configuration">Configuration</a> options can also be specified through the CLI.</p><p>We are working to make all the configuration options available in CLI.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="running-from-the-command-line"></a>Running from the command line<a aria-hidden="true" tabindex="-1" class="hash-link" href="#running-from-the-command-line" title="Direct link to heading">#</a></h2><p>Run all tests (default):</p><div class="mdxCodeBlock_1XEh"><div class="codeBlockContent_1u-d"><button type="button" aria-label="Copy code to clipboard" class="copyButton_10dd">Copy</button><div tabindex="0" class="prism-code language-bash codeBlock_3iAC"><div class="codeBlockLines_b7E3" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">corde</span></div></div></div></div></div><p>Run only the tests that were specified with a pattern or filename:</p><div class="mdxCodeBlock_1XEh"><div class="codeBlockContent_1u-d"><button type="button" aria-label="Copy code to clipboard" class="copyButton_10dd">Copy</button><div tabindex="0" class="prism-code language-bash codeBlock_3iAC"><div class="codeBlockLines_b7E3" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">corde path/to/test.js</span></div></div></div></div></div><div class="admonition admonition-info alert alert--info"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg></span>info</h5></div><div class="admonition-content"><p>You can run <code>corde --help</code> to view all available options.</p></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="available-commands"></a>Available Commands<a aria-hidden="true" tabindex="-1" class="hash-link" href="#available-commands" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="--config"></a><code>--config</code><a aria-hidden="true" tabindex="-1" class="hash-link" href="#--config" title="Direct link to heading">#</a></h3><p><strong>Alias: <code>-c</code></strong></p><p>Set config file path for the next execution. I.E:</p><div><ul role="tablist" aria-orientation="horizontal" class="tabs"><li role="tab" tabindex="0" aria-selected="true" class="tabs__item tabItem_1w39 tabs__item--active" style="outline:none">Long</li><li role="tab" tabindex="0" aria-selected="false" class="tabs__item tabItem_1w39" style="outline:none">Alias</li></ul><div role="tabpanel" class="margin-vert--md"><div><div class="mdxCodeBlock_1XEh"><div class="codeBlockContent_1u-d"><button type="button" aria-label="Copy code to clipboard" class="copyButton_10dd">Copy</button><div tabindex="0" class="prism-code language-bash codeBlock_3iAC"><div class="codeBlockLines_b7E3" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">corde --config </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain">/corde.js</span></div></div></div></div></div></div></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="--files"></a><code>--files</code><a aria-hidden="true" tabindex="-1" class="hash-link" href="#--files" title="Direct link to heading">#</a></h3><p><strong>Alias: <code>-f</code></strong></p><p>Set the path for all tests. Use this if you wan to specify a single path. for Array, use only <code>corde path1 path2</code></p><div><ul role="tablist" aria-orientation="horizontal" class="tabs"><li role="tab" tabindex="0" aria-selected="true" class="tabs__item tabItem_1w39 tabs__item--active" style="outline:none">Single Files</li><li role="tab" tabindex="0" aria-selected="false" class="tabs__item tabItem_1w39" style="outline:none">Multiple Files</li></ul><div role="tabpanel" class="margin-vert--md"><div><div class="mdxCodeBlock_1XEh"><div class="codeBlockContent_1u-d"><button type="button" aria-label="Copy code to clipboard" class="copyButton_10dd">Copy</button><div tabindex="0" class="prism-code language-bash codeBlock_3iAC"><div class="codeBlockLines_b7E3" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">corde --files ./tests/test1.js</span></div></div></div></div></div></div></div></div><p>Using alias:</p><div class="mdxCodeBlock_1XEh"><div class="codeBlockContent_1u-d"><button type="button" aria-label="Copy code to clipboard" class="copyButton_10dd">Copy</button><div tabindex="0" class="prism-code language-bash codeBlock_3iAC"><div class="codeBlockLines_b7E3" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">corde -f ./tests/test1.js ./tests/test2.js</span></div></div></div></div></div><div class="admonition admonition-note alert alert--secondary"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="14" height="16" viewBox="0 0 14 16"><path fill-rule="evenodd" d="M6.3 5.69a.942.942 0 0 1-.28-.7c0-.28.09-.52.28-.7.19-.18.42-.28.7-.28.28 0 .52.09.7.28.18.19.28.42.28.7 0 .28-.09.52-.28.7a1 1 0 0 1-.7.3c-.28 0-.52-.11-.7-.3zM8 7.99c-.02-.25-.11-.48-.31-.69-.2-.19-.42-.3-.69-.31H6c-.27.02-.48.13-.69.31-.2.2-.3.44-.31.69h1v3c.02.27.11.5.31.69.2.2.42.31.69.31h1c.27 0 .48-.11.69-.31.2-.19.3-.42.31-.69H8V7.98v.01zM7 2.3c-3.14 0-5.7 2.54-5.7 5.68 0 3.14 2.56 5.7 5.7 5.7s5.7-2.55 5.7-5.7c0-3.15-2.56-5.69-5.7-5.69v.01zM7 .98c3.86 0 7 3.14 7 7s-3.14 7-7 7-7-3.12-7-7 3.14-7 7-7z"></path></svg></span>note</h5></div><div class="admonition-content"><p>This is also a &quot;anonymous&quot; command. It means that you can just pass the files path without use <code>-f</code> or<br>
<code>--files</code>. I.E:</p><div class="mdxCodeBlock_1XEh"><div class="codeBlockContent_1u-d"><button type="button" aria-label="Copy code to clipboard" class="copyButton_10dd">Copy</button><div tabindex="0" class="prism-code language-bash codeBlock_3iAC"><div class="codeBlockLines_b7E3" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">corde ./tests/test1.js</span></div></div></div></div></div></div></div><hr><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="--init"></a><code>--init</code><a aria-hidden="true" tabindex="-1" class="hash-link" href="#--init" title="Direct link to heading">#</a></h3><p><strong>Alias: <code>-i</code></strong><br>
<strong>Default: <code>json</code></strong><br>
<strong>Options: <code>js json ts</code></strong></p><p>Initialize a config file. The available file formats are: <code>.js</code>, <code>.json</code>, <code>.ts</code>.</p><div><ul role="tablist" aria-orientation="horizontal" class="tabs"><li role="tab" tabindex="0" aria-selected="true" class="tabs__item tabItem_1w39 tabs__item--active" style="outline:none">Long</li><li role="tab" tabindex="0" aria-selected="false" class="tabs__item tabItem_1w39" style="outline:none">Alias</li></ul><div role="tabpanel" class="margin-vert--md"><div><div class="mdxCodeBlock_1XEh"><div class="codeBlockContent_1u-d"><button type="button" aria-label="Copy code to clipboard" class="copyButton_10dd">Copy</button><div tabindex="0" class="prism-code language-bash codeBlock_3iAC"><div class="codeBlockLines_b7E3" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">corde --init js</span></div></div></div></div></div></div></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="--validate"></a><code>--validate</code><a aria-hidden="true" tabindex="-1" class="hash-link" href="#--validate" title="Direct link to heading">#</a></h3><p><strong>Alias: <code>-v</code></strong></p><p>Search for corde configs and check if all required data is setted. Also check if file paths are valid.
<em>This does not check if tokens are valid.</em></p><div><ul role="tablist" aria-orientation="horizontal" class="tabs"><li role="tab" tabindex="0" aria-selected="true" class="tabs__item tabItem_1w39 tabs__item--active" style="outline:none">Long</li><li role="tab" tabindex="0" aria-selected="false" class="tabs__item tabItem_1w39" style="outline:none">Alias</li></ul><div role="tabpanel" class="margin-vert--md"><div><div class="mdxCodeBlock_1XEh"><div class="codeBlockContent_1u-d"><button type="button" aria-label="Copy code to clipboard" class="copyButton_10dd">Copy</button><div tabindex="0" class="prism-code language-bash codeBlock_3iAC"><div class="codeBlockLines_b7E3" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">corde --validate</span></div></div></div></div></div></div></div></div></div></article><div class="margin-vert--xl"><div class="row"><div class="col"><a href="https://github.com/lucasgmagalhaes/corde/blob/master/website/versioned_docs/version-2.0/CLI.mdx" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="1.2em" width="1.2em" preserveAspectRatio="xMidYMid meet" viewBox="0 0 40 40" style="margin-right:0.3em;vertical-align:sub"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div></div></div><div class="margin-vert--lg"><nav class="pagination-nav" aria-label="Blog list page navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/docs/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« Getting Started</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/docs/configurations"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Configuring Corde Â»</div></a></div></nav></div></div></div><div class="col col--3"><div class="tableOfContents_3SO_"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#running-from-the-command-line" class="table-of-contents__link">Running from the command line</a></li><li><a href="#available-commands" class="table-of-contents__link">Available Commands</a><ul><li><a href="#--config" class="table-of-contents__link"><code>--config</code></a></li><li><a href="#--files" class="table-of-contents__link"><code>--files</code></a></li><li><a href="#--init" class="table-of-contents__link"><code>--init</code></a></li><li><a href="#--validate" class="table-of-contents__link"><code>--validate</code></a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><h4 class="footer__title">Docs</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/docs/">Style Guide</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">Community</h4><ul class="footer__items"><li class="footer__item"><a href="https://stackoverflow.com/questions/tagged/corde" target="_blank" rel="noopener noreferrer" class="footer__link-item">Stack Overflow</a></li></ul></div></div><div class="text--center"><div>Corde 2020 </div></div></div></footer></div>
<script src="/styles.78702091.js"></script>
<script src="/runtime~main.c80c9b84.js"></script>
<script src="/main.a97df2ae.js"></script>
<script src="/common.570c17a0.js"></script>
<script src="/2.f7266800.js"></script>
<script src="/29.0b1f2317.js"></script>
<script src="/30.99c11695.js"></script>
<script src="/78311276.2010c58c.js"></script>
<script src="/17896441.364198ab.js"></script>
<script src="/e6aa8ecf.28a4f4e9.js"></script>
</body>
</html>