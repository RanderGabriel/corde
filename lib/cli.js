#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const clear_1 = __importDefault(require("clear"));
const figlet_1 = __importDefault(require("figlet"));
const commander_1 = __importDefault(require("commander"));
const initializer_1 = require("./initializer");
clear_1.default();
console.log(chalk_1.default.red(figlet_1.default.textSync('Concord', { horizontalLayout: 'full' })));
commander_1.default
    .version('0.0.1')
    .description("Discord bot testing framework")
    .option('-r, --run', 'run tests')
    .parse(process.argv);
if (commander_1.default.run) {
    initializer_1.begin();
}
if (!process.argv.slice(2).length) {
    commander_1.default.outputHelp();
}
//# sourceMappingURL=cli.js.map