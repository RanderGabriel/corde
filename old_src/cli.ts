#!/usr/bin/env node

import chalk from 'chalk';
import clear from 'clear';
import program from 'commander';
import figlet from 'figlet';
import pack from '../package.json';
import { execTestFiles } from './init';

clear();

console.log(chalk.red(figlet.textSync('corde', { horizontalLayout: 'full' })));

program
  .version(pack.version)
  .description('Discord bot testing framework')
  .option('-r, --run [file]', 'run tests')
  .parse(process.argv);

if (program.run) execTestFiles(program.run);
if (!process.argv.slice(2).length) program.outputHelp();
