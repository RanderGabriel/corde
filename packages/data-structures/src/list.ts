/**
 * List if an encapsulation of [Array](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array).
 * It's contains some others functions that are no present in Array itself
 * While provide access direct to main Array functions.
 *
 * The main goal is to reduce the acess to no relevant functions provided by Array and
 * specialize this class in iteration adition, reduction and search of elements.
 */
export class List<T> {
  private _data: T[];

  /**
   * Initialize a new List.
   */
  constructor() {
    this._data = [];
  }

  /**
   * Converts a array to List.
   * It do not modify the current values of the list.
   * @param data Array with values to be converted.
   * @returns A new instance of List. (The list has no references this instance).
   */
  public arrayToList(data: T[]): List<T> {
    const newList = new List<T>();
    for (const val of data) {
      newList.add(val);
    }
    return newList;
  }

  /**
   * Get the first elements that match with the predicate
   * does the same of [Array.find](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/find).
   * @param predicate Comparator to find the element.
   * @returns The first element of the list that match with the predicate,
   * or null.
   */
  public single(predicate: (value: T) => boolean): T {
    return this._data.find(predicate) ?? null;
  }

  /**
   * Same of
   * [Array.filter](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/filtro).
   * Finds a List of elements that match with a predicate.
   * @param predicate Comparator to find the elements.
   * @returns An list of all elements that match with the predicate.
   * returns an empty list if no element was found.
   */
  public filter(predicate: (value: T) => boolean): List<T> {
    const arr = this._data.filter(predicate);
    return this.arrayToList(arr);
  }

  /**
   * Picks the **first** element of the list, or null if no element exists.
   * @returns The first element or undefined if no element exists.
   */
  public first(): T | undefined {
    return this._data[0];
  }

  /**
   * Picks the **last** element of the list, or undefined if no element exists.
   * @returns The last element of the list of undefined if no element exists.
   */
  public last(): T | undefined {
    return this._data[this._data.length - 1];
  }

  /**
   * Removes all elements of the list.
   * @returns This instance.
   */
  public clear(): List<T> {
    this._data = [];
    return this;
  }

  /**
   * Copy all elements of the list to a new List.
   * @returns The cloned list.
   */
  public clone(): List<T> {
    return this.arrayToList(this._data.map((d) => d));
  }

  /**
   * Do a operation for each element of the list.
   * Same of [Array.foreach](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach).
   * @param operation Operation to execute in each element.
   * @returns this instance.
   */
  public each(operation: (val: T) => void): List<T> {
    for (const obj of this._data) {
      operation(obj);
    }
    return this;
  }

  /**
   * Select all properties informed in parameter that are contained in
   * the List object.
   * @param keys Properties of T to be picked.
   * @returns All picked properties of T in a new List.
   * Returns an empty list of no element was found.
   */
  public select<K extends keyof T>(...keys: K[]): List<Pick<T, K>> {
    const newList = new List<Pick<T, K>>();
    for (const obj of this._data) {
      const copy = {} as Pick<T, K>;
      keys.forEach((key) => (copy[key] = obj[key]));
      newList.add(copy);
    }
    return newList;
  }

  /**
   * Add a new element to list.
   * Note: undefined or null elements are not added.
   * @param values The new element.
   * @returns This instance.
   */
  public add(...values: T[]): List<T> {
    for (const val of values) {
      if (val) {
        this._data.push(val);
      }
    }
    return this;
  }

  /**
   * Checks if the list has at least one element.
   * @returns True if there are elements in list, false if don't.
   */
  public any() {
    return this._data.length > 0;
  }

  /**
   * Removes a element from the list.
   * @param data Element to be removed.
   * @returns This instance.
   */
  public remove(data: T): List<T>;
  /**
   * Removes an array of elements from the list.
   * @param data Array of elements.
   * @returns This instance.
   */
  public remove(data: T[]): List<T>;
  /**
   * Removes an list of elements from the list.
   * @param data List of elements.
   * @returns This instance.
   */
  public remove(data: List<T>): List<T>;
  public remove(data: T | T[] | List<T>) {
    if (!data) {
      return this;
    }
    if (this._isArray(data) || data instanceof List) {
      (data as T[]).forEach((d) => this._removeItem(d));
      return this;
    } else {
      this._removeItem(data);
      return this;
    }
  }

  /**
   * Get a element based on its **index**.
   * @param index Position in list of the element
   * @returns The element in the position informed.
   */
  public get(index: number): T;
  /**
   * Get a list of elements based in their **index**.
   * @param index Position of each element in list.
   * @returns A list with each element informed in array.
   * No existing elements will be ignored in adition on list.
   */
  public get(...index: number[]): List<T>;
  public get(index: number | number[]): T | List<T> {
    if (this._isArray(index)) {
      const values = new List<T>();
      for (const i of index) {
        const element = this._data[i];
        if (element) {
          values.add(element);
        }
      }
      return values;
    } else {
      return this._data[index];
    }
  }

  /**
   * Returns the index of the first element that matchs with the informed element.
   * Same of [Array.indexOf](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf).
   * @param element Element to have its index searched.
   * @returns Index of the searched element.
   */
  public indexOf(element: T): number {
    return this._data.indexOf(element);
  }

  public has(element: T) {
    return !!this._data.find((d) => d === element);
  }

  private _removeItem(data: T) {
    const index = this._data.indexOf(data, 0);
    if (index > -1) {
      this._data.slice(index, 1);
    }
  }

  private _isArray(content: any | any[]): content is any[] {
    return Array.isArray(content);
  }
}
