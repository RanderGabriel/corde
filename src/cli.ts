import arg from 'arg';
import { runTests } from './engine';

export function cli(args: string[]) {
  let files = parseArgumentsIntoOptions(args);
  if (!files) {
  } else {
    runTests(files);
  }
}

function parseArgumentsIntoOptions(rawArgs: string[]) {
  const args = arg(
    {},
    {
      argv: rawArgs.slice(2),
    },
  );
  return args._;
}
