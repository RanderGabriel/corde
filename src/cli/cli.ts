import { Command } from "commander";
import { go } from "./go";
import { init } from "./init";
import { validate } from "./validate";
import { configFileType } from "../types/types";
import { initErrorHandlers } from "../errorHandler";
import pack from "../package";
import { runtime } from "../common";
import { reader } from "../core";

initErrorHandlers();

export const program = new Command();

// Add basic information with default run all command
program
  .name("Corde")
  .usage("to start testings o corde [option] to use a specific command.")
  .description(pack.description)
  .version(`v${pack.version}`, "-v, --version");

program
  .option("-c, --config <type>", "Set config file path")
  .option(
    "-f, --files <path>",
    "Set the path for all tests. Use this if you wan to specify a single path." +
      " for Array, use only 'corde <path1> <path2>'",
  )
  .action(async (args) => {
    const options = program.opts();
    if (options.config) {
      runtime.configFilePath = options.config;
    }
    if (args) {
      runtime.testFiles = program.args;
    }
    if (options.files) {
      runtime.testFiles = options.files.split(" ");
    }
    await go();
  });

program
  .command("init [type]")
  .alias("i")
  .description("Initialize a config file with all possible options")
  .usage("[js ts json] or empty for default type (json)")
  .action((type: configFileType) => {
    init(type);
  });

program
  .command("validate")
  .alias("v")
  .description("Search for corde configs and check if all data are valid")
  .action(() => {
    const configs = reader.loadConfig();
    validate(configs);
    console.log("All configs are ok!");
  });

// tslint:disable-next-line: deprecation
if (process.env.ENV !== "TEST") {
  program.parse(process.argv);
} else {
  program.exitOverride();
}
