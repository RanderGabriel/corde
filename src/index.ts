/**
 * @file Automatically generated by barrelsby.
 */

 export * from "./cli";
export * from "./erros/index";
