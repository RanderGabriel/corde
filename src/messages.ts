export const messages = {
  TOTAL: " TOTAL: ",
  FAILURES: " FAILURES: ",
  SUCCESS: " SUCCESS: ",
  INFO: "INFO: ",
  EMPTY_TEST_FILE: "Empty test file",
  ALL_TESTS_PASSED: "All tests passed!",
  ALL_TESTS_FAIL: "All tests fail.",
  TESTS_PASSED_WITH_ERRORS: "Tests passed with errors.",
};
