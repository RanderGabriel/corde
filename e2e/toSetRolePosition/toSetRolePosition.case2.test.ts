import Utils from "../testUtils";
import { messages } from "../../src/messages";
import cli from "../cliRunner";
import { assert, spec } from "../pipeline";

spec("should fail", async () => {
  const command = Utils.buildCommandWithConfigPath("toSetRolePosition", "bot_case2.test.ts");
  const results = await cli.exec(command);
  assert(results.stdout).toContain(messages.ALL_TESTS_FAIL);
  assert(results.stdout).toContain(messages.FAILURES + " 1");
  assert(results.statusCode).toEqual(1);
});
