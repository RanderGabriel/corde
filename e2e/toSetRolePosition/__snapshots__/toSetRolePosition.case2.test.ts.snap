// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fail when trying to set role as mentionable 1`] = `
"
 FAIL   e2e\\\\toSetRolePosition\\\\__cordeTest__\\\\bot_case2.test.ts   100ms
   x <empty test name>   100ms
    
    expected: role position to change to 5
    received: position didn't change

    at toSetRolePosition e2e\\\\toSetRolePosition\\\\__cordeTest__\\\\bot_case2.test.ts:8:54
    at lib\\\\src\\\\clausures\\\\test.js:31:13
    at Queue.executeAsync (lib\\\\src\\\\data-structures\\\\queue.js:99:27)

Test Files: 1 failed. 1 total
Tests:      1 failed. 1 total
Time:       100ms
"
`;
